src/libraries/Common/tests/TestUtilities/System/Buffers/BoundedMemory.Unix.cs

```diff
@@ -118,29 +115,76 @@ public override void Unpin()
+
+                // Allocate number of pages to incorporate required (byteLength bytes of) memory and an additional page to create a poison page.
+                int pageSize = Environment.SystemPageSize;
+                int allocationSize = (int)(((byteLength / pageSize) + ((byteLength % pageSize) == 0 ? 0 : 1) + 1) * pageSize);
+                IntPtr buffer = mmap(IntPtr.Zero, (ulong)allocationSize, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+
+                if (buffer == IntPtr.Zero)
+                {
+                    throw new InvalidOperationException("Memory allocation failed.");
+                }
+
+                // Depending on the PoisonPagePlacement requirement (before/after) initialise the baseAddress and poisonPageAddress to point to the location
+                // in the buffer. Here the baseAddress points to the first valid allocation and poisonPageAddress points to the first invalid location.
+                // For PoisonPagePlacement.Before the first page is made inaccessible using mprotect and baseAddress points to the start of the second page.
+                // The allocation and protection is at the granularity of a page. Thus, `PoisonPagePlacement.Before` configuration has an additional accessible
+                // memory at the end of the page (bytes equivalent to `pageSize - (byteLength % pageSize)`).
+                // For `PoisonPagePlacement.After`, we adjust the baseAddress so that inaccessible memory is at the `byteLength` offset from the baseAddress.
+                IntPtr baseAddress = buffer + pageSize;
+                IntPtr poisonPageAddress = buffer;
+                if (placement == PoisonPagePlacement.After)
+                {
+                    baseAddress = buffer + (allocationSize - pageSize - byteLength);
+                    poisonPageAddress = buffer + (allocationSize - pageSize);
+                }
+
+                // Protect the page before/after based on the poison page placement.
+                if (mprotect(poisonPageAddress, (ulong) pageSize, PROT_NONE) == -1)
+                {
+                    throw new InvalidOperationException("Failed to mark page as a poison page using mprotect.");
+                }
+
+                AllocHGlobalHandle retVal = new AllocHGlobalHandle(buffer, (ulong)allocationSize);
+                retVal.SetHandle(baseAddress); // this base address would be used as the start of Span that is used during unit testing.
                 return retVal;
             }
 
-            // Do not provide a finalizer - SafeHandle's critical finalizer will
-            // call ReleaseHandle for you.
-
             public override bool IsInvalid => (handle == IntPtr.Zero);
 
             protected override bool ReleaseHandle()
             {
-                Marshal.FreeHGlobal(handle);
+                munmap(buffer, allocationSize);
                 return true;
```

Comment:
```suggestion
                return munmap(buffer, allocationSize) == 0;
```

It appears the Windows version does some error checking, should we instead follow the same pattern?